### PR TITLE
Support StreamConfiguration creation from JSON

### PR DESCRIPTION
**Pull request by senior**
This commit adds a public method to StreamConfiguration that allows hydrating a StreamConfiguration via a JSON String. Previously this was only package scoped and used in tests.

### CODE REVIEW

File: src/main/java/io/nats/client/api/StreamConfiguration.java [hunk 1]
```diff
@@ -143,9 +141,23 @@ static StreamConfiguration instance(JsonValue v) {
...
+     * Returns a StreamConfiguration deserialized from its JSON form.
+     *
+     * @see #toJson()
+     * @param json
+     * @return StreamConfiguration for the given json
+     * @throws JsonParseException
+     */
+    public static StreamConfiguration instance(String json) throws JsonParseException {
+        JsonValue parsedJson = JsonParser.parse(json);
+        return instance(parsedJson);
```
Comment (scottf) [inline src/main/java/io/nats/client/api/StreamConfiguration.java:154]: It's okay to combine these lines.

Comment (scottf) [inline src/main/java/io/nats/client/api/StreamConfiguration.java:154]: It's okay to combine these lines.

Comment (scottf) [inline src/main/java/io/nats/client/api/StreamConfiguration.java:154]: It's okay to combine these lines

File: src/main/java/io/nats/client/api/StreamConfiguration.java [hunk 2]
```diff
@@ -143,9 +141,23 @@ static StreamConfiguration instance(JsonValue v) {
...
+    }
+
     /**
      * Returns a JSON representation of this consumer configuration.
-     * 
+     *
+     * @see #instance(String)
```
Comment (scottf) [inline src/main/java/io/nats/client/api/StreamConfiguration.java:160]: Not sure this is necessary here.

File: src/test/java/io/nats/client/api/StreamConfigurationTests.java [hunk 3]
```diff
@@ -61,6 +66,87 @@ public void testRoundTrip() throws Exception {
...
+                add(ALLOW_ROLLUP_HDRS);
+                add(ALLOW_DIRECT);
+                add(MIRROR_DIRECT);
+                add(DENY_DELETE);
+                add(DENY_PURGE);
+                add(DISCARD_NEW_PER_SUBJECT);
+                add(METADATA);
+                add(FIRST_SEQ);
+            }
+        };
```
Comment (scottf) [inline src/test/java/io/nats/client/api/StreamConfigurationTests.java:123]: Very interesting voodoo here, I like.

File: src/test/java/io/nats/client/api/StreamConfigurationTests.java [hunk 4]
```diff
@@ -61,6 +66,87 @@ public void testRoundTrip() throws Exception {
...
+        String originalJson = ResourceUtils.dataAsString("StreamConfiguration.json");
+        JsonValue originalParsedJson = JsonParser.parse(originalJson);
+        originalParsedJson.map.put(NAME, new JsonValue("Inavlid*Name"));
+        assertThrows(IllegalArgumentException.class,
+                new Executable() {
+                    @Override
+                    public void execute() throws Throwable {
+                        StreamConfiguration.instance(originalParsedJson.toJson());
+                    }
+                });
```
Comment (scottf) [inline src/test/java/io/nats/client/api/StreamConfigurationTests.java:147]: make this a lambda
```
assertThrows(IllegalArgumentException.class, () -> StreamConfiguration.instance(originalParsedJson.toJson()));
```

File: src/test/java/io/nats/client/api/StreamConfigurationTests.java [hunk 5]
```diff
@@ -61,6 +66,87 @@ public void testRoundTrip() throws Exception {
...
+
+        String originalJson = ResourceUtils.dataAsString("StreamConfiguration.json");
+
+
+        for (String streamConfigFieldName: streamConfigFields) {
+            JsonValue originalParsedJson = JsonParser.parse(originalJson);
+            originalParsedJson.map.remove(streamConfigFieldName);
+            StreamConfiguration sc = StreamConfiguration.instance(originalParsedJson.toJson());
+            assertNotNull(sc);
+        }
```
Comment (scottf) [inline src/test/java/io/nats/client/api/StreamConfigurationTests.java:133]: can you add a comment at the top of this loop to describe what it's testing.

**Review by scottf — CHANGES_REQUESTED** [review]

File: src/main/java/io/nats/client/api/StreamConfiguration.java (hunk 1 above)
Comment (scottf) [reply src/main/java/io/nats/client/api/StreamConfiguration.java:154]:
```
return instance(JsonParser.parse(json));
```

File: src/test/java/io/nats/client/api/StreamConfigurationTests.java (hunk 4 above)
Comment (senior) [reply src/test/java/io/nats/client/api/StreamConfigurationTests.java:147]: I didn't know that was legal syntax for JDK8 syntax. I'll get it switched.

**Review by scottf — APPROVED** [review]
LGTM

**Pull request merged**